torch/testing/asserts.py

```diff
@@ -0,0 +1,186 @@
+    return _TestCase()._getDefaultRtolAndAtol(tensor1.dtype, tensor2.dtype)
+
+
+def _hide_internal_traceback(fn):
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        try:
+            return fn(*args, **kwargs)
+        except AssertionError:
+            for frame_info in inspect.trace():
+                frame_info.frame.f_locals.setdefault("__tracebackhide__", True)
+            raise
+
+    return wrapper
+
+
+# These attributes will be checked by default in every call to assert_tensors_(equal|allclose) and
+# assert_(equal|allclose) if the inputs are tensors.
+DEFAULT_CHECKED_ATTRIBUTES = (
+    "shape",
+    "dtype",
+)
+
+
+def _compare_tensors_meta(tensor1: torch.Tensor, tensor2: torch.Tensor, **attributes: bool) -> None:
+    if not (isinstance(tensor1, torch.Tensor) and isinstance(tensor2, torch.Tensor)):
+        raise AssertionError(f"Both inputs have to be tensors, but got {type(tensor1)} and {type(tensor2)} instead.")
+
+    for attr in DEFAULT_CHECKED_ATTRIBUTES:
+        attributes.setdefault(attr, True)
+
+    for attr, check in attributes.items():
+        if not check:
+            continue
+
+        try:
+            val1 = getattr(tensor1, attr)
+            val2 = getattr(tensor2, attr)
+            if val1 != val2:
+                raise AssertionError(f"The values for attribute '{attr}' do not match: {val1} != {val2}.")
+        except AttributeError as error:
+            raise UsageError(f"At least one of the inputs does not have an attribute '{attr}'.") from error
+
+
+def _compare_tensors(
+    tensor1: torch.Tensor,
+    tensor2: torch.Tensor,
+    asserter: Callable[[torch.Tensor, torch.Tensor], None],
+    **strict_attributes: bool,
+) -> None:
+    if any(tensor.dtype in (torch.complex32, torch.complex64, torch.complex128) for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for complex tensors is not supported yet.")
+    if any(tensor.is_quantized for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for quantized tensors is not supported yet.")
+    if any(tensor.is_sparse for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for sparse tensors is not supported yet.")
+
+    _compare_tensors_meta(tensor1, tensor2, **strict_attributes)
+
+    if tensor1.device != tensor2.device:
+        tensor1 = tensor1.cpu()
+        tensor2 = tensor2.cpu()
+
+    if tensor1.dtype != tensor2.dtype:
+        dtype = _get_comparison_dtype(tensor1, tensor2)
+        tensor1 = tensor1.to(dtype)
+        tensor2 = tensor2.to(dtype)
+
+    asserter(tensor1, tensor2)
+
+
+def _tensors_equal_asserter(tensor1: torch.Tensor, tensor2: torch.Tensor):
+    if not torch.all(torch.eq(tensor1, tensor2)):
+        raise AssertionError("ADDME")
+
+
+@_hide_internal_traceback
+def assert_tensors_equal(
+    tensor1: torch.Tensor, tensor2: torch.Tensor, *, warn_floating_point: bool = True, **strict_attributes: bool
```

Comment:
(a, b, /, *, ...) - make sure that the tensors being compared must be specified positionally